src/basic/terminal-util.c

```diff
@@ -228,49 +231,261 @@ int ask_char(char *ret, const char *replies, const char *fmt, ...) {
         }
 }
 
-int ask_string(char **ret, const char *text, ...) {
-        _cleanup_free_ char *line = NULL;
+typedef enum CompletionResult{
+        COMPLETION_ALREADY,       /* the input string is already complete */
+        COMPLETION_FULL,          /* completed the input string to be complete now */
+        COMPLETION_PARTIAL,       /* completed the input string so that is still incomplete */
+        COMPLETION_NONE,          /* found no matching completion */
+        _COMPLETION_RESULT_MAX,
+        _COMPLETION_RESULT_INVALID = -EINVAL,
+        _COMPLETION_RESULT_ERRNO_MAX = -ERRNO_MAX,
+} CompletionResult;
+
+static CompletionResult pick_completion(const char *string, char **completions, char **ret) {
+        _cleanup_free_ char *found = NULL;
+        bool partial = false;
+
+        STRV_FOREACH(c, completions) {
+
+                /* Ignore entries that are not actually completions */
+                if (!startswith(*c, strempty(string)))
+                        continue;
+
+                /* Store first completion that matches */
+                if (!found) {
+                        found = strdup(*c);
+                        if (!found)
+                                return -ENOMEM;
+
+                        continue;
+                }
+
+                /* If there's another completion that works truncate the one we already found by common
+                 * prefix */
+                size_t n = str_common_prefix(found, *c);
+                if (n == SIZE_MAX)
+                        continue;
+
+                found[n] = 0;
+                partial = true;
+        }
+
+        *ret = TAKE_PTR(found);
+
+        if (!*ret)
+                return COMPLETION_NONE;
+        if (partial)
+                return COMPLETION_PARTIAL;
+
+        return streq(strempty(string), *ret) ? COMPLETION_ALREADY : COMPLETION_FULL;
+}
+
+static void clear_by_backspace(size_t n) {
+        /* Erase the specified number of character cells backwards on the terminal */
+        for (size_t i = 0; i < n; i++)
+                fputs("\b \b", stdout);
+}
+
+int ask_string_full(
+                char **ret,
+                GetCompletionsCallback get_completions,
+                void *userdata,
+                const char *text, ...) {
+
         va_list ap;
         int r;
 
         assert(ret);
         assert(text);
 
+        /* Output the prompt */
         fputs(ansi_highlight(), stdout);
-
         va_start(ap, text);
         vprintf(text, ap);
         va_end(ap);
-
         fputs(ansi_normal(), stdout);
-
         fflush(stdout);
 
-        r = read_line(stdin, LONG_LINE_MAX, &line);
+        _cleanup_free_ char *string = NULL;
+        size_t n = 0;
+
+        /* Do interactive logic only if stdin + stdout are connected to the same place */
+        int fd_input = fileno(stdin);
+        int fd_output = fileno(stdout);
+        if (fd_input < 0 || fd_output < 0 || same_fd(fd_input, fd_output) <= 0)
```

Comment:
Hmm? STDIN/OUT_FILENO?

Reply:
yeah, that generally work, but it seemed a bit wrong to me, given that this function otherwise operates on FILE* and not raw fds. I mean, theoretically libc officially allows swapping out stdin/stdout by other FILE*, so that they don't point to STDIN_FILENO/STDOUT_FILENO, so I figured we might as well be accurate here.

I'll add a comment explaining this, I hope that's ok.